pkg/openstack/instances.go

```diff
@@ -622,17 +634,47 @@ func nodeAddresses(srv *servers.Server, ports []ports.Port, networkingOpts Netwo
 	}
 
 	// process the rest
-	type Address struct {
-		IPType string `mapstructure:"OS-EXT-IPS:type"`
-		Addr   string
-	}
-
 	var addresses map[string][]Address
 	err := mapstructure.Decode(srv.Addresses, &addresses)
 	if err != nil {
 		return nil, err
 	}
 
+	// Get the private networks that are not directly connected
+	if len(addresses) < len(allPrivates) {
+		extraPrivates := make(map[string][]Address)
+		// For each private network
+		for network, interfaceAddresses := range allPrivates {
+			found := false
+			// For each address in the private network
+			for _, interfaceAddress := range interfaceAddresses {
+				// Check if the address is directly assigned
+				for _, srvAddresses := range addresses {
+					for _, srvAddress := range srvAddresses {
+						if interfaceAddress.Addr == srvAddress.Addr {
```

Comment:
quite many for loops here, perhaps worth of using sub function(s) instead. It makes it more readable and easier to test if needed.

you could make function like `func findInterfaceAddresses(interfaceAddresses, addresses) bool`

Reply:
Good point, will change.